Promote Options::name to pretty_type::name utility

knelli2 opened the pull request:
## Proposed changes

I've been finding myself using `Options::name<T>()` more and more outside of the context of options now that more structs have a `name()` member. This PR promotes `Options::name<T>()` to the `pretty_type::name<T>()` utility.

Also, some uses of `pretty_type::short_name<T>()` have been replaced with `pretty_type::name` where a struct might want to use it's `name()` member over the name of the struct itself.

<!--
At a high level, describe what this PR does.
-->

### Upgrade instructions

<!--
If this PR makes changes that other people should be aware of when upgrading
their code, describe what they should do between the two UPGRADE INSTRUCTIONS
lines below.
-->
<!-- UPGRADE INSTRUCTIONS -->

<!-- UPGRADE INSTRUCTIONS -->

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `new feature` if appropriate.

### Further comments

<!--
If this is a relatively large or complex change, kick off the discussion by
explaining why you chose the solution you did and what alternatives you
considered, etc...
-->

knelli2 committed:
Remove Options::name()

and replace with pretty_type::name().

knelli2 committed:
Add pretty_type::name<T>()

Identical to Options::name(). Returns the result of the `name()`
member of a class. If a class does not have a `name()` member,
return the result of `pretty_type::short_name<T>()` instead.

knelli2 committed:
Replace short_name with name in some instances

Don't use it in places where we actually want the short_name of the
struct.

See docs of pretty_type::name<T>() for when not to use it.

nilsvu commented on tests/Unit/Utilities/Test_PrettyType.cpp:
```diff
@@ -249,190 +272,204 @@ SPECTRE_TEST_CASE("Unit.Utilities.PrettyType.short_name", "[Utilities][Unit]") {
+  test_name_func<ShortName>();
 }
+
+SPECTRE_TEST_CASE("Unit.Utilities.PrettyType.name", "[Utilities][Unit]") {
```
Please combine the two test cases into one. Test cases have a nonnegligible startup time.

nilsvu commented on src/Options/OptionsDetails.hpp:
```diff
@@ -92,7 +92,7 @@ using options_in_group =
 template <typename T>
 struct yaml_type {
-  static std::string value() { return pretty_type::short_name<T>(); }
+  static std::string value() { return pretty_type::name<T>(); }
```
In the commit message you say that sometimes it would be better to use `short_name` still, "for example in tags". Why so?

knelli2 replied:
If you have 
```
struct MyTag {
  static std::string name() { return pretty_type::name<MyTag>(); }
};
```

`pretty_type::name` will end up in a recursive call to the tag's `name()` function. Because of compile time stuff, I'm not sure if this is actually recursive or if it ends, but when getting the name of a struct from within the `name()` member itself I felt it better to use `short_name` instead to avoid this all together.

nilsvu replied:
In these cases, shouldn't we just delete the `name` function and make sure the consuming code uses `pretty_type::name`?

knelli2 replied:
Yes I suppose I can change `DataBox/TagName.hpp` to use `pretty_type::name` now and remove the name member from tags that just have `short_name<Tag>` implemented. Still doesn't prevent somebody in the future from making a new tag that uses `pretty_type::name` inside the `name()` member though...

kidder replied:
DataBox tags currently should only have a `name` function defined if `name` returns something other than `pretty_type::short_name<Tag>`   There are `TestHelpers::db::test_TAGTYPE_tag` functions that check this if the tags are properly tested.  So if you change the behavior, please update the test helpers...

kidder replied:
and if you find untested tags, please test them...

nilsvu replied:
My suggestion is to warn not to use `pretty_type::name` inside a `name` function if this can lead to recursion. That's fairly obvious, but why not add a warning. We don't have to deal with changing any DataBox / tags things in this PR, but if you want to make a decision then I'm not stopping you :)

nilsdeppe replied:
Sorry, what do you mean by "add a warning"? Do you mean in doxygen or somehow in the code? If the latter I'm not sure how to implement it, so providing some detail would probably be useful.

knelli2 replied:
I didn't realize there was already a check that the `name()` member of a tag must return something different than `pretty_type::short_name<Tag>()`. This alleviates my concerns there so I don't think I need to change db tag names. I'll add a warning to not use `pretty_type::name` inside a name function like @nilsvu suggested.

nilsvu replied:
@nilsdeppe I meant in docs

nilsvu commented on src/Utilities/PrettyType.hpp:
```diff
@@ -705,4 +706,26 @@ template <typename T>
+ * doesn't have a `name()` member, call `pretty_type::short_name<T>()` instead.
+ */
+template <typename T>
+std::string name() {
```
What do you think about moving this into a shorter namespace such as `tt` now that it is used so much throughout the code, and to avoid confusion with `short_name? Should be a fairly straightforward find-replace right? Get a second reviewer's opinion before doing it.

nilsdeppe replied:
I don't understand how this `name()` function is a type trait. Why would it make sense to move a non-type trait thing into the type trait namespace? I think `pretty_type` is the right place for this function

wthrowe replied:
I agree with @nilsdeppe.

nilsvu reviewed (changes requested):
LGTM, thanks for doing this 👍

nilsdeppe commented:
@wthrowe since this changes option parsing code would you be willing to at least do a brief review of it?

wthrowe reviewed (changes requested):
Requesting changes because I want to see the answers to other reviewers' questions.

knelli2 commented:
Fixed conflicts and squashed changes directly. Only changes were combining tests in `Test_PrettyType.cpp` and adding a warning to `pretty_type::name` docs. Also clarified commit message of second commit.

nilsvu commented on src/Utilities/PrettyType.hpp:
```diff
@@ -705,4 +706,31 @@ template <typename T>
+ * \warning Do not use this inside the `name()` member of a DataBox tag. This
+ * can lead to recursion as `pretty_type::name<Tag>()` will call the `name()`
+ * member of the tag. A DataBox tag should only specify a `name()` member if the
+ * name is different than `pretty_type::short_name<Tag>()`
```
I think remove the mention of tags. The recursion issue is independent of DataBox things. Just warn of the possible recursion when using this inside a name function.

knelli2 replied:
Done! :)
